zipline/gens/sim_engine.pyx

```diff
@@ -59,28 +64,50 @@ cdef class MinuteSimulationClock:
 
     @cython.boundscheck(False)
     @cython.wraparound(False)
-    cdef dict calc_minutes_by_day(self):
-        cdef dict minutes_by_day
-        cdef int day_idx
-        cdef object day
+    cdef dict calc_minutes_by_session(self):
+        cdef dict minutes_by_session
+        cdef int session_idx
+        cdef object session
 
-        minutes_by_day = {}
-        for day_idx, day in enumerate(self.trading_days):
-            minutes_by_day[day] = pd.to_datetime(
-                self.market_minutes(day_idx), utc=True, box=True)
-        return minutes_by_day
+        minutes_by_session = {}
+        for session_idx, session in enumerate(self.sessions):
+            minutes_by_session[session] = pd.to_datetime(
+                self.market_minutes(session_idx), utc=True, box=True)
+        return minutes_by_session
 
     def __iter__(self):
         minute_emission = self.minute_emission
 
-        for day in self.trading_days:
-            yield day, DAY_START
+        for idx, session in enumerate(self.sessions):
+            yield session, SESSION_START
+
+            bts_minute = self.before_trading_start_minutes[idx]
+            regular_minutes = self.minutes_by_session[session]
+
+            # we have to search anew every session, because there is no
+            # guarantee that any two session start on the same minute
+            bts_idx = regular_minutes.searchsorted(bts_minute)
+
+            if bts_idx == len(regular_minutes):
+                # before_trading_start is after the last close, so don't emit
+                # it
+                for minute in regular_minutes:
+                    yield minute, BAR
+                    if minute_emission:
+                        yield minute, MINUTE_END
+            else:
+                # emit all the minutes before bts_minute
+                for minute in regular_minutes[0:bts_idx]:
+                    yield minute, BAR
+                    if minute_emission:
+                        yield minute, MINUTE_END
 
-            minutes = self.minutes_by_day[day]
+                yield bts_minute, BEFORE_TRADING_START_BAR
 
-            for minute in minutes:
-                yield minute, BAR
-                if minute_emission:
-                    yield minute, MINUTE_END
+                # emit all the minutes after bts_minute
+                for minute in regular_minutes[bts_idx:]:
+                    yield minute, BAR
+                    if minute_emission:
+                        yield minute, MINUTE_END
```

Comment:
Instead of repeating this 3 times over different iterables, would it make sense to factor out a function?